### PR TITLE
cli: bookmark rename: add `--overwrite-existing` flag

### PR DESCRIPTION
**Pull request by gechr**
I tried to use `jj bookmark rename [--force] old new` today as a shortcut for `jj bookmark delete` followed by `jj bookmark move` only to discover it wasn't implemented.

Add `--overwrite-existing` to `jj bookmark rename` as a convenient way to delete/move in one step.

### CODE REVIEW

**Comment by joyously** [comment]
The `-f` flag usually means "from", so I don't think it should be used for "force".

**Comment by gechr** [comment]
@joyously Ah yeah, good point. Totally slipped my mind. Old git habits die hard...

I've dropped the `-f` entirely.

**Comment by martinvonz** [comment]
We also generally avoid `--force` and instead name it something that explains in what way it's forcing. Perhaps `jj bookmark rename --overwrite-existing`?

**Comment by yuja** [comment]
Perhaps, we'll also need to "overwrite" existing tracked state of remote bookmarks? The current implementation only handles "New -> Tracked" transition because the `new_bookmark` name doesn't exist locally.

**Comment by gechr** [comment]
@yuja What does `jj bookmark set` do in this case?

Plus, I think the following warning covers the scenario you're describing?

https://github.com/jj-vcs/jj/blob/8a3d94683b9eeb413b7cfcdb2508aa1ea755f28d/cli/src/commands/bookmark/rename.rs#L154-L168

**Comment by yuja** [comment]
> What does `jj bookmark set` do in this case?

`jj bookmark set`/`move` isn't the command to change the association of local and remote bookmarks. There are no "old"/"new" bookmarks pair.

https://github.com/jj-vcs/jj/blob/8a3d94683b9eeb413b7cfcdb2508aa1ea755f28d/cli/src/commands/bookmark/rename.rs#L154-L168

It is the warning for "Tracked -> Tracked" transition, in which case, we cannot copy the old tracked state (= the last merged revision) to the new name.

**Comment by gechr** [comment]
@yuja This ended up getting pretty hairy and I had to reach to an LLM for assistance. Please take a look at my most recent push and let me know if it's what you had in mind!

**Comment by yuja** [comment]
The renamed state should be overwritten from the old bookmarks (if possible), so something like this? Perhaps, `jj bookmark rename --overwrite-existing` would be equivalent to `jj bookmark forget new && jj bookmark rename`.

| old       | original new      | renamed new |
| --------- | ----------------- | ----------- |
| untracked | absent untracked  | untracked   |
| untracked | absent tracked    | untracked   |
| untracked | present untracked | untracked   |
| untracked | present tracked   | untracked + warning? |
| tracked   | absent untracked  | tracked     |
| tracked   | absent tracked    | tracked |
| tracked   | present untracked | warning (= untracked) |
| tracked   | present tracked   | untracked + warning OR warning (= tracked) |

Note: "absent tracked" will become "absent untracked" once the original "new" local bookmark gets deleted. We might have to do `set_local_bookmark_target(new_bookmark, RefTarget::absent())` or untrack absent-tracked remotes manually.

> It is the warning for "Tracked -> Tracked" transition

This comment was wrong. It's the warning for "Tracked/Untracked -> Tracked" transition.

That's indeed complicated.

File: cli/src/commands/bookmark/rename.rs [hunk 1]
```diff
@@ -92,19 +92,42 @@ pub fn cmd_bookmark_rename(
...
+                // absent+tracked remotes are cleaned up by deleting the local
+                // bookmark below
+                None
+            } else {
+                Some(symbol.remote.to_owned())
             }
-            !remote_ref.is_tracked()
         })
-        .map(|(symbol, _)| symbol.remote.to_owned())
         .collect::<HashSet<_>>();
```
Comment (yuja) [inline cli/src/commands/bookmark/rename.rs]: nit: Rather than building `HashSet`s of various conditions, I think it's simpler to query the original view.

```rust
let base_repo = workspace_command.repo().clone();
let base_view = base_repo.view();
...
let existing_ref = base_view.get_remote_bookmark(new_remote_bookmark);
if existing_ref.is_present() && !existing_ref.is_tracked() { // existing_untracked_remotes.contains(new_remote_bookmark.remote)
```

This will also stabilize the order of `existing_present_tracked_remotes` warnings.

File: cli/src/commands/bookmark/rename.rs [hunk 2]
```diff
@@ -92,19 +92,42 @@ pub fn cmd_bookmark_rename(
...
+    if args.overwrite_existing {
+        // Delete the overwritten bookmark first. This cleans up any
+        // absent+tracked remote refs via set_local_bookmark_target.
+        tx.repo_mut()
+            .set_local_bookmark_target(new_bookmark, RefTarget::absent());
+        // Untrack present+tracked remotes of the overwritten bookmark.
+        for remote in &existing_present_tracked_remotes {
+            tx.repo_mut()
+                .untrack_remote_bookmark(new_bookmark.to_remote_symbol(remote));
+        }
```
Comment (yuja) [inline cli/src/commands/bookmark/rename.rs:95]: nit: perhaps, we can do "untrack" for all remotes? Then we wouldn't have to do `.set_local_bookmark_target(..., absent)` to untrack absent remote refs.

File: cli/src/commands/bookmark/rename.rs [hunk 3]
```diff
@@ -124,9 +147,35 @@ pub fn cmd_bookmark_rename(
...
+            writeln!(
+                ui.warning_default(),
+                "Tracking of remote bookmark {new_bookmark}@{remote} was dropped.",
+                new_bookmark = new_bookmark.as_symbol(),
+                remote = remote.as_symbol(),
+            )?;
+            writeln!(
+                ui.hint_default(),
+                "Use `jj bookmark track` to re-track if needed.",
+            )?;
```
Comment (yuja) [inline cli/src/commands/bookmark/rename.rs:157]: Can you add a few more tests that cover this and some other edge cases (if any)?

I think it's also interesting to run "overwrite" renames in colocated repository, where `@git` remote bookmarks exist.

File: cli/src/commands/bookmark/rename.rs [hunk 4]
```diff
@@ -62,53 +64,51 @@ pub fn cmd_bookmark_rename(
...
+        base_view.remote_bookmarks_matching(&StringMatcher::exact(old_bookmark), &remote_matcher)
+    {
+        if !remote_ref.is_tracked() {
+            continue;
+        }
+        if remote_ref.is_present() {
+            tracked_present_old_remotes_exist = true;
+        }
+        let new_remote_bookmark = new_bookmark.to_remote_symbol(symbol.remote);
+        let existing_ref = base_view.get_remote_bookmark(new_remote_bookmark);
```
Comment (yuja) [inline cli/src/commands/bookmark/rename.rs]: Perhaps, we should use the current `tx.repo().view()` here because existing "new" tracked bookmarks would have been untracked above. Then, we can remove `if existing_ref.is_present() && existing_ref.is_tracked() { continue; }` since re-tracking tracked bookmark is noop.

File: cli/tests/test_bookmark_command.rs [hunk 5]
```diff
@@ -680,7 +689,78 @@ fn test_bookmark_rename() {
     [EOF]
     ");
```
Comment (yuja) [inline cli/tests/test_bookmark_command.rs:710]: It's also interesting to see `bookmark rename <OLD> <OLD>` with/without `--overwrite-existing`. Without `--overwrite-existing`, it should be an error. With overwrite, maybe the result is noop?

Comment (gechr) [reply cli/tests/test_bookmark_command.rs:710]: Nice catch, with `--overwrite-existing` it actually deleted the old bookmark and set the new bookmark but the delete happened after the set, so it ended up deleting the bookmark.

I've pushed a change to ensure `--overwrite-existing` is a noop and without `--overwrite-existing` is an error.
